test/sync_integration_test.js

```diff
@@ -0,0 +1,292 @@
+      bob.disconnect()
+      charlie.disconnect()
+
+      // console.log('---------------- making changes')
+
+      // each one makes a change
+      alice.change(s => (s.alice = 1))
+      bob.change(s => (s.bob = 1))
+      charlie.change(s => (s.charlie = 1))
+
+      // while they're disconnected, they have divergent docs
+      assert.notDeepStrictEqual(alice.doc, bob.doc)
+      assert.notDeepStrictEqual(bob.doc, charlie.doc)
+      assert.notDeepStrictEqual(alice.doc, charlie.doc)
+
+      // console.log('---------------- reconnecting')
+
+      alice.connect()
+      bob.connect()
+      charlie.connect()
+
+      // after connecting, their docs converge
+      assert.deepStrictEqual(alice.doc, bob.doc)
+      assert.deepStrictEqual(bob.doc, charlie.doc)
+      assert.deepStrictEqual(alice.doc, charlie.doc)
+    })
+  })
+})
+
+function connect(a, b) {
+  const channel = new Channel()
+  a.connectTo(b.peerId, channel)
+  b.connectTo(a.peerId, channel)
+}
+
+class ConnectedDoc extends EventEmitter {
+  constructor(peerId, doc) {
+    super()
+    this.peerId = peerId
+    this.doc = A.clone(doc)
+    this.peers = {}
+  }
+
+  connectTo(remotePeerId, channel) {
+    this.connected = true
+    const peer = new Peer(this.peerId, remotePeerId, this.doc, channel)
+    this.peers[remotePeerId] = peer
+    peer.update(this.doc)
+    peer.on('change', doc => {
+      // console.log(`doc changed ${this.peerId}<-${remotePeerId}`, doc)
+      this.update(doc, remotePeerId)
+    })
+  }
+
+  update(doc, remotePeerId) {
+    this.doc = doc
+    for (const peerId in this.peers) {
+      this.peers[peerId].update(doc)
+    }
+  }
+
+  change(fn) {
+    const updatedDoc = A.change(this.doc, fn)
+    this.update(updatedDoc)
+  }
+
+  disconnect() {
+    this.connected = false
+    for (const peerId in this.peers) {
+      this.peers[peerId].channel.leave()
+    }
+  }
+
+  connect() {
+    this.connected = true
+    for (const peerId in this.peers) {
+      const peer = this.peers[peerId]
+      peer.channel.join()
+      peer.update(this.doc)
+    }
+  }
+}
+
+class Peer extends EventEmitter {
+  constructor(peerId, remotePeerId, doc, channel) {
+    super()
+    this.syncState = A.initSyncState()
+    this.peerId = peerId
+    this.remotePeerId = remotePeerId
+    this.doc = doc
+
+    this.updates = 0
+
+    channel.join()
+    this.channel = channel
+    channel.addListener('data', (senderPeerId, msg) => {
+      if (senderPeerId === this.peerId) return // ignore our own messages
+      this.receive(msg)
+    })
+  }
+
+  update(doc) {
+    const [syncState, msg] = A.generateSyncMessage(doc, this.syncState)
+    this.doc = doc
+    this.syncState = syncState
+    this.send(msg)
+  }
+
+  receive(msg) {
+    const [doc, syncState] = A.receiveSyncMessage(this.doc, this.syncState, msg)
+    // console.log(`received update ${this.peerId}<-${this.remotePeerId}`, doc)
+    this.doc = doc
+    this.syncState = syncState
+    this.emit('change', doc)
+  }
```

Comment:
Maybe I am tracing the code incorrectly here, but I don't see you running update after receiving a message? Instead it looks like after receiving a message, you emit a change message that causes _other_ peers in a ConnectedDoc to run their update calls. Thats not what you want. You want:
1. receive a sync message and apply it
2. now check if you have a sync message to send (generateSyncMessage) with the new sync state
3. send it if there is a message to send